Added provision for users to more accurately change their gps location

ashwinkjoseph opened the pull request:
### Issue Reference
Fixes #125 

### Summarize
I have added a button below for users to change their GPS Location. 
When pressed, an OpenStreetMap is rendered with the help of OpenLayers 5.1

The user is given the provision to search for a position or to drag the marker to point user's location more accurately.

after user has made changes to gps location, the change is reflected in the form when user clicks on change button

ranjithn commented:
Clicking on "change gps location" more than once, adds a new map to the page.

Probably due to calling createMap() again in this snippet:

		changeGPSLocationBtn.addEventListener('click', function(event){
			document.getElementById('gpslocationchooser').hidden = false;
			createMap();
		});

answerquest commented:
@ashwinkjoseph Suggestion: instead of user having to pin/drag a marker AND pan/zoom the map, how about doing like ola/uber and just moving the map, keeping a crosshairs fixed on center, and taking map-center co-ords as the location? I was at #225 and was about to start work on this, but I can only do standalone leaflet, can't understand how to merge with all the existing code which it seems you've done.

yeaske commented:
@msrshahrukh100 Doesn't the google API have rate limiting? Isn't #266 a more sustainable solution?

msrshahrukh100 commented:
well it has some rate limiting, so basically, i'll have to pay for the extra requests. If #266 is a more sustainable solution than please go ahead with it, as it's the people that matter more not a pull request.

gladiatorvarg commented:
@msrshahrukh100 do you think we are getting that many requests...

ashwinkjoseph commented:
@ranjithn @answerquest  working on solving the issues you pointed out

bobinson commented:
@ ashwinkjoseph - can you update on #225 too ? @naveenpf  was about to merge this.

ashwinkjoseph commented:
@ranjithn I have fixed the issue you have mentioned, 
@bobinson  yea I'll look into that now

ashwinkjoseph commented:
@bobinson  did you mean the suggestion made by @answerquest ? if so yes, Im working on it now

ashwinkjoseph commented:
I'll fix this conflict in the styles.css along with the change that @answerquest requested

ashwinkjoseph commented:
I added the feature requested by @answerquest  and have resolved the conflict in styles.css

bobinson commented:
@ashwinkjoseph  - can you get this PR tested ?

bobinson commented:
https://keralarescue.slack.com/archives/CCASHFV2B/p1534490133000100 refer this @ashwinkjoseph

ashwinkjoseph commented:
@bobinson I have tested it on my local machine

drtechie commented on mainapp/templates/mainapp/request_form.html:
```diff
@@ -29,19 +41,17 @@ <h3 id="myModalLabel">Allow Location - നിങ്ങളുടെ സ്ഥല
+	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
+	<script src="https://cdn.jsdelivr.net/npm/lodash@4.17.10/lodash.min.js" integrity="sha256-/GKyJ0BQJD8c8UYgf7ziBrs/QgcikS7Fv/SaArgBcEI=" crossorigin="anonymous"></script>
+	<!-- <script src="http://www.openlayers.org/api/OpenLayers.js"></script> -->
+	<script src="https://cdn.rawgit.com/openlayers/openlayers.github.io/master/en/v5.1.3/build/ol.js"></script>
```
Consider loading this later or only if user clicks the button. Defer loading this JS.

bobinson commented:
check the readem of the repo for slack join link @ashwinkjoseph

drtechie commented on mainapp/templates/mainapp/request_form.html:
```diff
@@ -29,19 +41,17 @@ <h3 id="myModalLabel">Allow Location - നിങ്ങളുടെ സ്ഥല
-  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
+	</div>
+	<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
+	<script src="https://cdn.jsdelivr.net/npm/lodash@4.17.10/lodash.min.js" integrity="sha256-/GKyJ0BQJD8c8UYgf7ziBrs/QgcikS7Fv/SaArgBcEI=" crossorigin="anonymous"></script>
```
We need not load a 32KB lodash package to use one function. I am seeing only throttle being used.

drtechie commented:
Functionality wise this looks good. 

Can you add a feedback after 'Change' clicked? I clicked it multiple times without knowing whats happening. Also adding a line explaining what the buttons are about will be good. We can change the "Change GPS location' to 'Find' or something like that. The double 'Change' will also cause some confusion.

justusrk commented:
I have also tested this on a phone from my local server. Works fine.

justusrk commented:
"Pick your Location" ?

vrthra commented:
Verified it works on desktop; Agree with @drtechie about the change button.

drtechie commented:
Tested on desktop and its working fine. The location popup also appears. 
Test picking location mobile and that is working fine too.
Mobile location popup didn't come - I think its related to the https issue - not sure.

answerquest commented:
> I have not been added to your slack team and so I cant refer to the archive you commented

Join here @ashwinkjoseph 
 https://join.slack.com/t/keralarescue/shared_invite/enQtNDE4NzUyNjg4MjQ3LTJiMDU0ZmFhODNlNDE3ZDc4ZmFlMGI0YmQ0MzI0NWYyNThlOTgwYTM2Y2JlYzMxMDMxMzUwY2E2MmVmNDQyNmE

ashwinkjoseph commented:
@drtechie @justusrk  working on the changes you have mentioned

ashwinkjoseph commented:
@answerquest thanks , I have joined now

vrthra commented:
I checked in the mobile (moto-g). The location picker is not responsive, the pin does not move where you point with the touch.

vrthra commented:
@ashwinkjoseph what is your slack id?

ashwinkjoseph commented:
I have made the modifications according to your suggestions:
1. Loading the map library only when user clicks on the button and added a loader for better UX
2. Added the alert for when user successfullty changes location
3. Have imported only the necessary functions from lodash (the code depended on module but since I am not that used to using a bundler with something like django, I haent done that, please feel free to optimize that part)

gladiatorvarg commented:
@ashwinkjoseph I have made some changes to your code. you can look at it [here](https://github.com/gladiatorvarg/rescuekerala). Would you be interested in merging it with your code?

ashwinkjoseph commented:
@gladiatorvarg  im looking at the code

gladiatorvarg commented:
@ashwinkjoseph sent PR to your repo

sumodm commented:
@ashwinkjoseph: Functionality working on Chrome, not on firefox, 

1. UI is a little confusing. Next to change location you have submit button, so one would assume that submit would submit location. But that is below as 'change'. Also it is not obvious that one has to press change. 
 
Suggestion
i. Move submit to top right.
ii. Show 'Change' initially in red once user presses search and then change to green when user clicks change and show the button as saved.  

2. The pin is still located in a different place than where the mouse pointer is clicked. Would be great if you can fix that.

3. Doesn't seem to work on firefox (
![screenshot from 2018-08-17 16-28-27](https://user-images.githubusercontent.com/1060369/44262960-a4b61780-a23a-11e8-85a0-62432f86069b.png)


4. _**nice to have**_  It would be great if you can have autosuggest names when typed location is typed in.

@vrthra One potential problem is that OSM is not as rich. One of you would have to make the call on this.

ashwinkjoseph commented:
@sumodm  okay, I'm working on it

ashwinkjoseph commented:
@sumodm how about changing submit button text to submit request and changing "change" to submit location?

ashwinkjoseph commented:
@sumodm which version of firefox were you using, It was working for me in Firefox Developer Edition

rabeesh commented:
@ashwinkjoseph  

**Desktop:**

Tested in chrome 68, Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36
Working fine.

**Mobile:** 
chrome 68.0.3440.91
android 6.0.1

But it's not working in:
chrome 50.0.2661.89
Android 4.4.2

I think we can skip that old mobile version case.

rabeesh commented:
It would be nice to move  "change GPS Location" input near to "GPS Coordinates - GPS നിർദ്ദേശാങ്കങ്ങൾ"

drtechie commented:
Tested this on desktop and mobile.
Both chrome and Firefox are loading the popup when accessed via https.

ashwinkjoseph commented:
Im halting my work on 266 as 289 is being merged, you can let me know if I need to be working on this

gladiatorvarg commented:
I created a new PR with some code refactoring. Please continue testing here - https://github.com/IEEEKeralaSection/rescuekerala/pull/479.

biswaz commented:
@ashwinkjoseph Please continue work, now collaborate with @gladiatorvarg. We might need this. Reviewer said this had responsive problems on mobile.
